### DISCUSSION TITLE
Dependency Inheritance and Overriding Best Practice

### DISCUSSION THREAD

**Discussion by TicTocCroc** [pointfreeco/swift-composable-architecture#1891] [Q&A] [answered]
Suppose I have a parent reducer that contains or navigates to a child reducer, and the parent reducer's body looks something like this:

```Swift
var body: some ReducerProtocol<State, Action> {
  Reduce { state, action in
      // … parent reducer logic…
  }
  Scope(state: \.childState, action: /Action.child) {
      Child()
  }
}
```

Suppose each reducer also has the same dependency, for example:

```Swift
struct CountClient {
  var getCount: () -> Int
}
// …
// Parent reducer dependency
@Dependency(\.countClient) var countClient
// …
// Child reducer dependency
@Dependency(\.countClient) var countClient
```

In order to override this dependency, in say a preview, I can use the `withDependencies()` function when I'm constructing a parent store:

```Swift
let store: StoreOf<Parent> = {
  let parent = withDependencies {
    $0.countClient = CountClient(fetch: { 100 })
  } operation: {
      Parent()
  }
  return StoreOf<Parent>(initialState: Parent.State(), reducer: parent)
}()
```

This will correctly override the `CountClient` dependency on the parent, but the child reducer does **not** use this overridden dependency, because when the child reducer is instantiated in the parent's body, it's outside the scope of the `withDependencies()` call.

To make overrides propagate to the child reducer, should the parent's body be something like this?

```Swift
var body: some ReducerProtocol<State, Action> {
  Reduce { state, action in
      // … parent reducer logic…
  }
  Scope(state: \.childState, action: /Action.child) {
      Child().dependency(\.countClient, self.countClient)
  }
}
```

This seems to work, but it means explicitly setting the child's dependency in the parent's body, which seems clunky. Furthermore, if you wanted child reducers to always inherit overridden dependencies from the parent, you'll be configuring your dependencies this way for most of your app's reducer.

Finally, if the parent and child don't share a dependency, then it doesn't seem possible to override the child's dependencies at all when instantiating the parent.

This makes me think I'm missing something when it comes to overriding dependencies on a hierarchy of reducers. Or maybe I'm running into bugs, or have configured my app incorrectly.

I can provide a sample project demonstrating some of this, if it's helpful.

**Comment by mbrandonw** [comment, marked as answer]
Hi @TicTocCroc, I believe there are two ways you can fix this. Rather than wrapping the creation of the `Parent` reducer in `withDependencies`, you should be using the `.dependency` reducer operator in the preview:

```swift
static var previews: some View {
  ParentView(
    store: Store(
      initialState: Parent.State(),
      reducer: Parent()
        .dependency(\.countClient, …)
    )
  )
}
```

And another option is the `Store` initializer takes a trailing closure that allows you to override dependencies for the entire runtime:

```swift
static var previews: some View {
  ParentView(
    store: Store(
      initialState: Parent.State(),
      reducer: Parent()
    ) {
      $0.countClient = …
    }
  )
}
```

If neither of those options work, then perhaps a sample project would be helpful.

**Reply by tgrapperon** [reply]
I was writing an answer, but Brandon beat me to it! As an aside, you don't need to manually inject the dependency from the parent into `Child` when doing like this, as dependencies from "parents" are automatically forwarded to child reducers, like `Environment` values are propagated into child views in SwiftUI.

**Reply by TicTocCroc** [reply]
Both of those solutions work, thank you! I guess I was using the wrong touch point for overriding dependencies.

The current [documentation for dependencies](https://pointfreeco.github.io/swift-composable-architecture/0.50.0/documentation/composablearchitecture/dependencymanagement) only mentions overriding dependencies in the context I showed above (as a "particular reducer inside a larger composed reducer"), and refers the reader to the swiftui-dependencies documentation for further information, which mainly focuses on the `withDependencies()` mechanisms for overriding.

This is why I focused on these methods.

It might be helpful to emphasize that a "top" level reducer can override in the same way, with "children" inheriting the override, as well mentioning the store initializer for overriding dependencies.

**Reply by mbrandonw** [reply]
Hi @TicTocCroc, that's great feedback. I think we are leaning too much on Dependencies' docs for TCA, and we should provide more TCA-specific docs. We will get on that soon. Thanks!
